src/Modules/Tags.elm

```diff
@@ -0,0 +1,352 @@
+import Json.Decode as Decode
+import Json.Decode.Pipeline as Decode
+import Shared.Styles as Styles
+import Task
+
+
+type alias Model =
+    { tags : Dict Int { name : String, value : String }
+    , newTagName : String
+    , newTagValue : String
+    , isInherited : Bool
+    , newTagError : Maybe String
+    , error : Dict Int String
+    }
+
+
+type alias Tag =
+    { name : String
+    , value : String
+    }
+
+
+decoder : Decode.Decoder Tag
+decoder =
+    Decode.succeed Tag
+        |> Decode.required "name" Decode.string
+        |> Decode.optional "value" Decode.string ""
+
+
+initialModel : { tags : Dict String String, isInherited : Bool } -> Model
+initialModel { tags, isInherited } =
+    { tags =
+        tags
+            |> Dict.toList
+            |> List.indexedMap
+                (\index value ->
+                    ( index
+                    , { name = Tuple.first value
+                      , value = Tuple.second value
+                      }
+                    )
+                )
+            |> Dict.fromList
+    , newTagName = ""
+    , newTagValue = ""
+    , isInherited = isInherited
+    , newTagError = Nothing
+    , error = Dict.empty
+    }
+
+
+type Msg
+    = NoOp
+    | NameChanged Int { name : String }
+    | ValueChanged Int { value : String }
+    | Removed Int
+    | TagAdded
+    | NewTagNameChanged String
+    | NewTagValueChanged String
+    | TagValidated Int
+    | Submitted
+    | Closed
+
+
+update : Msg -> Model -> { onSubmit : Model -> msg, onClose : msg, onFocus : msg } -> ( Model, Cmd msg )
+update msg model config =
+    case msg of
+        NoOp ->
+            ( model, Cmd.none )
+
+        NameChanged key { name } ->
+            ( { model
+                | tags =
+                    Dict.update
+                        key
+                        (Maybe.map (\a -> { name = name, value = a.value }))
+                        model.tags
+              }
+            , Cmd.none
+            )
+
+        ValueChanged key { value } ->
+            ( { model
+                | tags =
+                    Dict.update
+                        key
+                        (Maybe.map (\a -> { name = a.name, value = value }))
+                        model.tags
+              }
+            , Cmd.none
+            )
+
+        Removed key ->
+            ( { model
+                | tags = Dict.remove key model.tags
+                , error = Dict.remove key model.error
+              }
+            , Cmd.none
+            )
+
+        TagAdded ->
+            let
+                tagAlreadyExists : Bool
+                tagAlreadyExists =
+                    model.tags
+                        |> Dict.toList
+                        |> List.map (Tuple.second >> .name)
+                        |> List.filter ((==) model.newTagName)
+                        |> List.head
+                        |> Maybe.map (\_ -> True)
+                        |> Maybe.withDefault False
```

Comment:
This could have been a single `Dict.foldl` call 🤷🏻

Reply:
`Dict.Extra.any` =|

Reply:
`model.tags |> Dict.values |> List.any ((==) model.newTagName)` if you don't want to use Dict.Extra.